lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -655,17 +640,22 @@ SymbolicValue ConstExprFunctionCache::computeConstantValue(SILValue value) {
   if (auto *builtin = dyn_cast<BuiltinInst>(value))
     return computeConstantValueBuiltin(builtin);
 
-  if (auto *apply = dyn_cast<ApplyInst>(value))
-    if (computeCallResult(apply) == ResultCode::Success) {
-      assert(calculatedValues.count(apply));
-      return calculatedValues[apply];
-    }
+  if (auto *apply = dyn_cast<ApplyInst>(value)) {
+    auto callResult = computeCallResult(apply);
+
+    // If this failed, return the error code.
+    if (callResult.hasValue())
```

Comment:
`computeCallResult` sets `calculatedValues[apply]` in all cases, so this failure-handler is redundant.

Opportunity for future improvement, I don't think it's necessary in this PR:

The return value from `computeCallResult` is always redundant. Making the return value `void` would simplify `computeCallResult` and reduce opportunities for bugs (e.g. accidentally return a `SymbolicValue` on success, or a caller accidentally thinks that `None` is a failure).

I also think that the other functions returning `Optional<SymbolicValue>` have slightly confusing signature, because `None` is usually used to indicate some kind of failure. I didn't look closely enough to see if those are also redundant -- if they are not redundant, I think it would be nice to change the signature to something that makes it a bit more obvious that `None` is the success case, like `Optional<UnknownReason>`.

Reply:
There's still small value in the return because you wouldn't have to do a hash table lookup right after calling `computeCallResult`.

Reply:
Also, calls can have "only indirect" results, which means they won't have a normal result.